### ISSUE TITLE
Let's discuss the license

### ISSUE THREAD

**Issue by ChALkeR** [qmlweb/qmlweb#52]
This is aimed to replace #12.

Current issues:
 1. The LICENSE is a modified BSD-3-Clause, we should pick an SPDX-approved variant.
 2. The LICENSE is a BSD-3-Clause, but the files in `src/qtcore/qml/lib/` have BSD-2-Clause headers.
 3. We bundle `uglify-js` which is `BSD-2-Clause`.
 4. Most of web js libraries just use MIT (see #12 for previous discussion).
 5. Given the 1, 2 and 3, I don't know what to put into the `license` field of `package.json`, and `npm` spits on me with these messages:

 ```
npm WARN qmlweb@0.0.4 license should be a valid SPDX license expression
 ```

I guess everyone should be fine with either of these three licenses, but I must ask everyone.

We can also discuss the preferred license here. After that, I will make a PR with the chosen license that everyone would need to review again.

Author | [BSD-3-Clause](https://spdx.org/licenses/BSD-3-Clause.html) | [BSD-2-Clause](https://spdx.org/licenses/BSD-2-Clause.html) | [MIT](https://spdx.org/licenses/MIT.html) | Preferred | Notes
------------ | ------------- | ------------- | ------------- | ------------- | ------------- | -------------
@lpaimen | + | + | + | No preference |  |
@akreuzkamp | + | + | + | BSD-2-Clause (with parser included), BSD-2-Clause or MIT (with parser excluded) |  |
@jangmarker | + | + | + |  | One commit, through KDE reviewboard |
@Somsubhra | + | + | + | No preference |  |
@JoshuaKolden | + | + | + | MIT | Only Readme changes |
@ChALkeR | + | + | + | MIT or BSD-2-Clause |  |
@Plaristote |  | + | + | MIT? |  |
@labsin | + | + | + | No preference |  |
@arnopaehler | + | + | + |  |  |
@pavelvasev | + | + | + | No preference |  |
@henrikrudstrom | + | + | + | No preference |  |
@aspidites | + | + | + | MIT | Only Readme changes |
@shry15harsh | + | + | + | No preference | One commit, through KDE reviewboard |
@igosoft | + | + | + |MIT or BSD-2-Clause |  |
@aspotashev | + | + | + | | One commit, through KDE Phabricator |
@stephenmdangelo | + | + | + | | |

Please leave a comment about yourself — are you fine with:
 1. Relicensing all your present and future contributions to this repo and present contributions to other public qmlweb repos and forks under the [BSD-3-Clause](https://spdx.org/licenses/BSD-3-Clause.html) license.
 2. Relicensing all your present and future contributions to this repo and present contributions to other public qmlweb repos and forks under the [BSD-2-Clause](https://spdx.org/licenses/BSD-2-Clause.html) license.
 3. Relicensing all your present and future contributions to this repo and present contributions to other public qmlweb repos and forks under the [MIT](https://spdx.org/licenses/MIT.html) license.
 4. Which license would you prefer?
 5. Perhaps some arguments for the preferred license, to keep the discussion going.

**Comment by akreuzkamp** [comment]
I'm fine with all three licenses and hereby grant permission to use my code under BSD-2, BSD-3 or MIT at your choice.
I would probably go for BSD-2, at least as long as we depend on uglify for simplicity. I'm not sure, if we can at all license the whole thing as MIT, as long as we use BSD-2 licensed code, can we?

**Comment by ChALkeR** [comment]
@akreuzkamp Actually, that makes sense, and perhaps BSD-2-Clause would be a better choice because of this.

**Comment by labsin** [comment]
I'm fine with all three licenses and hereby grant permission to use my code under BSD-2, BSD-3 or MIT at your choice.
No preference.

**Comment by henrikrudstrom** [comment]
no strong opinions here either, i'd be fine if you flipped a coin

**Comment by Plaristote** [comment]
I'm no expert, so I'm fine with both licenses. If uglify limits our choice, we can still think about this matter if we get rid of uglify as a dependency.
We may even want to make it optional (we don't always need to parse javascript: users of gulp-qmlweb won't need it).

**Comment by ChALkeR** [comment]
@henrikrudstrom Does that mean that you are fine with all 3 of them?

@Plaristote 
> I'm fine with both licenses. 

There are three of them. Which ones did you mean?

> We may even want to make it optional (we don't always need to parse javascript: users of gulp-qmlweb won't need it).

Yes, see #2 — if we split into separate modules, perhaps the modules could have different licenses. Not sure yet.

**Comment by henrikrudstrom** [comment]
im ok with all of them

**Comment by Plaristote** [comment]
Sorry, I meant MIT and BSD-2 licenses.

**Comment by ChALkeR** [comment]
@Plaristote Like this? *Updates the table.*

**Comment by Somsubhra** [comment]
I am fine with all three licenses with no specific preferences.

**Comment by Plaristote** [comment]
@ChALkeR Yes, that's it.

**Comment by ChALkeR** [comment]
@Plaristote @Somsubhra @henrikrudstrom @akreuzkamp @labsin @igosoft, thanks!

**Comment by henrikrudstrom** [comment]
no expert on licensing, but noticed a discussion on MPL license on the qml-web project. it looked quite interesting : [here is an article about it](http://www.gatsbymedia.net/development-articles/why-we-use-the-mpl)

**Comment by ChALkeR** [comment]
Here is [MPL](https://spdx.org/licenses/MPL-2.0.html). I would be fine with it, but I don't see any actual benefits for the project over the simple licenses listed here. I'm also not an expert on licensing, though.

**Comment by lpaimen** [comment]
I'm fine with all of these.

I remember quickly evaluating the licenses and picking BSD due the third clause (no promotion with my name) but anyways I don't see that happening (in a negative spirit). I don't have any preference over licenses.

**Comment by ChALkeR** [comment]
@lpaimen, thanks!

**Comment by henrikrudstrom** [comment]
From what i understood MPL includes file level copyleft, so its quite permissive while being copyleft, but thats not really a concern for me, as long as its permissive.

**Comment by aspidites** [comment]
I'm fine with all three though I prefer MIT for its permissibility. That said, feel free to relicense my contributions under either of the BSD variants if that is the preferred license of other contributors.

**Comment by shry15harsh** [comment]
I'm fine with all three licenses.
No preference for any license.

**Comment by ChALkeR** [comment]
@aspidites @shry15harsh Thanks!

**Comment by pavelvasev** [comment]
Which one will allow using qmlweb in commercial projects? With closed sources? Free?

**Comment by ChALkeR** [comment]
@pavelvasev I suppose that all of them do.

**Comment by pavelvasev** [comment]
@ChALkeR Thank you!

**Comment by ChALkeR** [comment]
Ping @jangmarker, @JoshuaKolden, @arnopaehler, @pavelvasev.

**Comment by ChALkeR** [comment]
@igosoft Btw, you updated the table, but could I ask you to also leave an explicit comment here (like the others did), please?

**Comment by JoshuaKolden** [comment]
tl;dr: I vote MIT.

The BSD 3 and 4 clause licenses are for high profile individuals and
organizations who release a project to open source.  The additional clauses
are related to protecting the brand, or 'name' of the original creator.  I
don't think these issues apply to our group.

The question then is MIT or BSD 2 Clause license.

MIT licensed source can be included safely in almost any other project
(while retaining the MIT notice at the top of the files), because the MIT
license is *more* permissive.  This is good for propagation of MIT licensed
projects.  Some companies are restricted from using BSD, GPL and others
because they force the commercial software to become open source under most
conditions.

Because the MIT license is more permissive there are some licenses that are
incompatible due mainly to not allowing it's use in closed source projects.

*MIT License:*

*Pro:* Easer to accept, better uptake.  Makes QMLweb easier to use.

*Con:* Can't use source from some open source projects as part of QMLweb.

*BSD 2 clause License:*

*Pro:* Easier to include source from other projects.

*Con:* Excludes QMLweb from use in some projects particularly commercial
closed source projects.

MIT will also require more diligence on our part since we have to be
careful not to knowingly include source from non-permisive OS Licenses.

Nevertheless I prefer wider use over forcing open source, so I vote MIT
license.

j

**Comment by JoshuaKolden** [comment]
Also, as a rule, I prefer not to bundle were possible, since that forces
duplication and version collisions in the deployment context .  However,
open source projects under different licensing should be includable as
stand alone tools, source files, or libraries without effecting the QMLweb
code or license if downstream users of QMLweb can 'opt out' (not use,
vendorize, pre-compile, buy closed source license, etc.) from the tools
provided by the other code.

The prohibition is only against using the source code of less permissive
licensed code directly (i.e. copy/paste).

In fewer words: MIT license might be a bit of a hassle, but it isn't all
that bad in practice.

**Comment by Plaristote** [comment]
Right now, the uglify parser is the only thing we're bundling, and there's a `browserified` branch in which I've removed this bundle by using browerisfy.

The source inclusion won't be an issue once we're a true nodejs package, and since we're aiming that for before 1.0.0, the cons of MIT don't really apply to us, do they ?

So that seems to be another win for the MIT license.

**Comment by pavelvasev** [comment]
@JoshuaKolden wikipedia says that BSD is [ok with commercial](https://en.wikipedia.org/wiki/BSD_licenses#Commercial_license_compatibility)..?

> Some companies are restricted from using BSD

Why?.. Or is it just standard practice?..

**Comment by pavelvasev** [comment]
Can we suggest 2 licences, BSD and commercial MIT-like, and sell commercial on perpetual free update basis and use the income for example for meetings or conferences? Or is there another practice in common world for that, e.g. licence all for MIT and gain "funding"?

**Comment by ChALkeR** [comment]
@pavelvasev I don't think that selling licenses is a great idea, nor do I think that any of the proposed licenses (including the ones that's already there, though those are mixed up) impose any significant restrictions on anyone using qmlweb in a closed-source project. There would be no reason for anyone to pay for removing BSD/MIT restrictions, because there bacially aren't any considerable restrictions =)

**Comment by pavelvasev** [comment]
@ChALkeR @JoshuaKolden tells that BSD brings some problem for commercial use.. (waiting for Joshua answer).

**Comment by JoshuaKolden** [comment]
@pavelvasev I may have implied too much by lumping BSD with GPL in that sentence. I meant that more as a comment on what is possible with MIT. Companies can make blanket assertions about MIT license source with no ongoing obligations outside of retaining the MIT copyright notice with the code.  This means only software developers at a company need be involved in the decision.  This is not true of BSD 2-clause because it has the ongoing specific performance obligation that the license be included even in "binary" distributions.  

    "Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution." 

This constitutes a public facing legal exposure, and may involve other departments such as design, packaging and marketing.  This still isn't that much of an obligation, certainly nothing like copyleft, but my point was MIT is an easy "yes" whereas almost everything else gets put in the "requires consideration" bucket.

See also:

1. http://programmers.stackexchange.com/questions/121998/mit-vs-bsd-vs-dual-license
2. http://www.tomhull.com/ocston/docs/mozgpl.html

**Comment by pavelvasev** [comment]
Dear @JoshuaKolden , thank you very much for the greatest explanation! ;-)

**Comment by jangmarker** [comment]
I'm fine with all three licenses.

**Comment by ChALkeR** [comment]
Thanks all.

There is no uglifyjs here anymore, but [/src/qtcore/qml/lib/parser.js](https://github.com/qmlweb/qmlweb/blob/master/src/qtcore/qml/lib/parser.js) is based on it, so it would better to keep that file BSD-2-Clause licensed.

Note that parser.js would be moved to a separate repo soon (after the merges would be complete), and qmlweb-parser would be an _optional_ component (because it's possible to use the runtime without the parser, e.g. when everything is pre-processed and served in an already parsed form from the server, see gulp-qmlweb).

That would free the code here from uglifyjs-derived code and we would be able to choose any license everyone here agrees on.

**Comment by ChALkeR** [comment]
@JoshuaKolden, I see that you stated that you prefer MIT from those, but could you please also explicitly state which ones are you fine with?

@akreuzkamp You are the only one who gave preference to BSD-2, hence I'm asking you this: is the dependance on uglifyjs the only reason? If yes, would moving the parser that is based on uglifyjs to a separate repo as an optional dependency solve that question? More specific: taking the parser out, what license would you prefer for the qmlweb runtime and why?

**Comment by akreuzkamp** [comment]
Yes, uglifyjs' license is the only reason. :)

**Comment by akreuzkamp** [comment]
But I've got another reason against BSD-3, which is [KDE's licensing policy](https://community.kde.org/Policies/Licensing_Policy), that allows BSD as well as MIT licenses, but states 
> Ensure that the BSD license does not contain the so called 'advertisement clause'.

If we ever want to become/stay a KDE project (#26) we should stick to their policy.

**Comment by ChALkeR** [comment]
@akreuzkamp Thanks!

**Comment by JoshuaKolden** [comment]
Aside from my preference I'm fine with any of the three.

**Comment by ChALkeR** [comment]
@JoshuaKolden Thanks for clarification! :smile:

**Comment by ChALkeR** [comment]
I just sent a letter to @arnopaehler.

**Comment by arnopaehler** [comment]
I am ok with all three licenses and my code may be distributed under any one of them.

**Comment by ChALkeR** [comment]
@arnopaehler Thanks!

**Comment by ChALkeR** [comment]
Added @aspotashev to the list, see #206.

**Comment by aspotashev** [comment]
I agree with relicensing my code in qmlweb under any of these 3 licenses (BSD-3-Clause, BSD-2-Clause or MIT).

**Comment by ChALkeR** [comment]
@akreuzkamp Thanks!

**Comment by ChALkeR** [comment]
@aspotashev Thanks!

**Comment by stephenmdangelo** [comment]
I am good with any of the licenses.

**Comment by ChALkeR** [comment]
@stephenmdangelo Thanks!

**Comment by ChALkeR** [comment]
Closed in 04fa2d79967672b9a9893fe13fbf91898d12526f, thanks all!
